hashconsing_derive/src/lib.rs

```diff
@@ -0,0 +1,168 @@
+                                            }]),
+                                        },
+                                    }
+                                    .into()
+                                },
+                                FnArg::Typed(PatType {
+                                    attrs: Vec::new(),
+                                    pat: Box::new(Pat::Ident(PatIdent {
+                                        attrs: Vec::new(),
+                                        by_ref: None,
+                                        mutability: None,
+                                        ident: id,
+                                        subpat: None,
+                                    })),
+                                    colon_token: Token![:](Span::call_site()),
+                                    ty: Box::new(f.ty.clone()),
+                                }),
+                            )
+                        })
+                        .unzip();
+
+                    let calling_args = Punctuated::from_iter(arg_names);
+
+                    let variant_name = Expr::Path(ExprPath {
+                        attrs: Vec::new(),
+                        qself: None,
+                        path: Path {
+                            leading_colon: None,
+                            segments: Punctuated::from_iter(vec![
+                                PathSegment {
+                                    ident: ident.clone(),
+                                    arguments: PathArguments::None,
+                                },
+                                PathSegment {
+                                    ident: v.ident.clone(),
+                                    arguments: PathArguments::None,
+                                },
+                            ]),
+                        },
+                    });
+
+                    let variant_expr = if calling_args.is_empty() {
+                        variant_name
+                    } else {
+                        ExprCall {
+                            attrs: Vec::new(),
+                            func: Box::new(variant_name),
+                            paren_token: Paren(Span::call_site()),
+                            args: calling_args,
+                        }
+                        .into()
+                    };
+
+                    (Punctuated::from_iter(arg_types), variant_expr)
+                })
+                .unzip();
+
+            quote! {
+                consign! {
+                    let #factory_name = consign(50) for #ident ;
+                }
+
+                impl #struct_name {
+                    #(pub fn #variant_names(#variant_field_function_args) -> Self {
+                        Self(#factory_name.mk(#variant_field_calling_args))
+                    })*
+                }
+            }
+        }
+        _ => todo!("Doesn't yet support things that aren't enums for impl"),
```

Comment:
Would be better to generate an idiomatic proc-macro error, using `compile_error!` or [proc_macro_error](https://docs.rs/proc-macro-error/latest/proc_macro_error/).

Message-wise, I would expect something like `unsupported syntax: hashconsing expected an enum definition` or something similar.

Reply:
Sure